torch/csrc/jit/codegen/cuda/test/test_gpu.cpp

```diff
@@ -23741,6 +23784,116 @@ TEST_F(NVFuserTest, FusionIssue1770Repro_CUDA) {
+  auto tv1 = makeSymbolicTensor(1);
+  fusion->addInput(tv1);
+
+  auto tv2 = add(tv0, tv1);
+
+  auto tv3 = sin(tv2);
+  auto tv4 = cos(tv2);
+
+  fusion->addOutput(tv3);
+  fusion->addOutput(tv4);
+
+  tv2->split(0, 10);
+
+  struct Selector : public MaxInfoSpanningTree::Selector {
+    TensorView* tv0;
+    TensorView* tv3;
+    virtual bool allowPasC(TensorView* from, TensorView* to) override {
+      return to == tv0;
+    }
+    virtual bool allowCasP(TensorView* from, TensorView* to) override {
+      return to == tv3;
+    }
+    Selector(TensorView* tv0, TensorView* tv3) : tv0(tv0), tv3(tv3) {}
+  } selector(tv0, tv3);
+
+  TransformPropagator propagator(tv2);
+  MaxRootDomainInfoSpanningTree(tv2, &selector).traverse(&propagator);
+
+  TORCH_CHECK(tv0->nDims() == 2);
+  TORCH_CHECK(tv1->nDims() == 1);
+  TORCH_CHECK(tv2->nDims() == 2);
+  TORCH_CHECK(tv3->nDims() == 2);
+  TORCH_CHECK(tv4->nDims() == 1);
+}
+
+TEST_F(NVFuserTest, FusionTransormPropagatorPos_CUDA) {
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  auto tv0 = makeConcreteTensor({22, 105});
+  fusion->addInput(tv0);
+
+  auto tv1 = sin(tv0);
+  fusion->addOutput(tv1);
+
+  tv1->split(0, 2);
+  tv1->split(-1, 3);
+  tv1->split(-1, 5);
+
+  TransformPropagator propagator(tv1, 2);
+  MaxRootDomainInfoSpanningTree(tv1, 2).traverse(&propagator);
+
+  TORCH_CHECK(tv0->nDims() == 3);
+  TORCH_CHECK(tv0->axis(0)->extent()->evaluateInt() == 11);
+  TORCH_CHECK(tv0->axis(1)->extent()->evaluateInt() == 2);
+  TORCH_CHECK(tv0->axis(2)->extent()->evaluateInt() == 105);
+}
+
+TEST_F(NVFuserTest, FusionMaxRootDomainInfoSpanningTreePrintTwice_CUDA) {
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  auto tv0 = makeSymbolicTensor(3);
+  fusion->addInput(tv0);
+
+  auto tv1 = sum(tv0, {0});
+  auto tv2 = neg(tv1);
+
+  fusion->addOutput(tv2);
+
+  tv1->split(0, 10);
+
+  struct Printer : public MaxInfoSpanningTree::Propagator {
+    std::stringstream ss;
+    virtual void propagateTvPasC(TensorView* from, TensorView* to) override {
+      ss << "propagateTvPasC" << std::endl;
+      ss << "from: " << from << std::endl;
+      ss << "to: " << to << std::endl;
+    }
+    virtual void propagateTvCasP(TensorView* from, TensorView* to) override {
+      ss << "propagateTvCasP" << std::endl;
+      ss << "from: " << from << std::endl;
+      ss << "to: " << to << std::endl;
+    }
+  } printer1, printer2;
+  printer1.ss << std::endl;
+  printer2.ss << std::endl;
+
+  MaxRootDomainInfoSpanningTree path(tv1);
+  path.traverse(&printer1);
+  path.traverse(&printer2);
+
+  auto expect = R"ESCAPE(
+propagateTvPasC
+from: T1_l[ rS8{( ceilDiv(i1, 10) )}, rS9{10}, iS4{i2}, iS5{i3} ]
+to: T0_g[ iS0{i1}, iS1{i2}, iS2{i3} ]
+propagateTvCasP
+from: T1_l[ rS8{( ceilDiv(i1, 10) )}, rS9{10}, iS4{i2}, iS5{i3} ]
+to: T2_g[ iS6{i2}, iS7{i3} ]
+)ESCAPE";
+  TORCH_CHECK(printer1.ss.str() == expect);
+  TORCH_CHECK(printer2.ss.str() == expect);
+}
```

Comment:
Please see this as an example of adding a custom propagator, and traversing the same path multiple times with a different propagator.